### PR TITLE
Add missing impl blocks for item reexported from private mod in JSON output

### LINKED ISSUE #102583: rustdoc JSON: struct fn missing if re-exported from private mod
### Step-by-step

1. Put this in `repro.rs`:
```rust
mod private_mod {
    pub struct S;

    impl S {
        pub fn missing_from_rustdoc_json() {}
    }
}

pub use private_mod::*;
```

2. Run `rustdoc +nightly-2022-10-02 repro.rs -Zunstable-options -wjson`

3. Check if `missing_from_rustdoc_json` is in the generated JSON: `grep missing_from_rustdoc_json doc/repro.json`

### Actual result

Item is missing.

### Expected result

Item is present.

### Remarks

The item is present in the HTML version of the docs (`rustdoc +nightly-2022-10-02 repro.rs  && open doc/repro/index.html`), or in the JSON if `private_mod` is made `pub`.

@rustbot labels +A-rustdoc-json +T-rustdoc +C-bug

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #102583.

Since we don't inline for the JSON output, the impl blocks from private modules are not present when we generate the output. To go around this limitation, in case the impl block doesn't have `#[doc(hidden)]` and is implementing a public item, we don't strip it.

cc @fmease @aDotInTheVoid 
r? @notriddle

### CODE REVIEW

File: src/test/rustdoc-json/reexport_method_from_private_module.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Regression test for <https://github.com/rust-lang/rust/issues/102583>.
```
Comment (aDotInTheVoid) [inline src/test/rustdoc-json/reexport_method_from_private_module.rs:1]: Nit: This should be in `src/test/rustdoc-json/reexport.rs`

File: src/test/rustdoc-json/reexport_method_from_private_module.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+// @has "$.index[*][?(@.name=='missing_from_rustdoc_json')]"
```
Comment (aDotInTheVoid) [inline src/test/rustdoc-json/reexport_method_from_private_module.rs:3]: Nit: Could you also check that S has an impl block, and the impl block has just the expected method

File: src/test/rustdoc-json/reexport_method_from_private_module.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// @!has "$.index[*][?(@.name=='hidden_impl')]"
+// @!has "$.index[*][?(@.name=='hidden_fn')]"
+
+#![no_std]
+
+mod private_mod {
+    pub struct S;
+
+    impl S {
+        pub fn missing_from_rustdoc_json() {}
```
Comment (aDotInTheVoid) [inline src/test/rustdoc-json/reexport_method_from_private_module.rs:13]: Nit: I don't think this name makes sense now that the issue is fixed.

File: src/librustdoc/passes/stripper.rs [hunk 4]
```diff
@@ -151,6 +153,23 @@ pub(crate) struct ImplStripper<'a> {
     pub(crate) document_private: bool,
 }
 
+impl<'a> ImplStripper<'a> {
+    #[inline]
```
Comment (aDotInTheVoid) [inline src/librustdoc/passes/stripper.rs:157]: Why is this needed? Is their a policy for when to use #[inline], esp given this isn't used across crates

**Review by aDotInTheVoid — COMMENTED** [review]
Looks good, just some small nits around the test

File: src/librustdoc/passes/stripper.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/passes/stripper.rs:157]: It's just a habit of mine: when a function is used often and less than 15 lines, I put this attribute.

File: src/test/rustdoc-json/reexport_method_from_private_module.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs:1]: :+1:

File: src/test/rustdoc-json/reexport_method_from_private_module.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs:13]: :+1:

File: src/test/rustdoc-json/reexport_method_from_private_module.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs:3]: So, for whatever reason I can't do:

```
// @has "$.index[*][?(@.kind=='impl' && @.inner.for.inner.name=='S')]"
```

but this one works:

```
// @has "$.index[*][?(@.kind=='impl')].inner.for.inner.name" '"S"'
```

I wanted to use it to use to first count the number of impl blocks for `S` and then count the number of items in this impl. But since it's not possible, I'll just leave it as is I guess...

**Comment by GuillaumeGomez** [comment]
Updated! I'm a bit sad I couldn't extend the test as I wanted but I guess it's enough for now. If you find a way to do it, I'd be very interested!

File: src/test/rustdoc-json/reexport_method_from_private_module.rs (hunk 2 above)
Comment (aDotInTheVoid) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs]:
```
// @has "$.index[*][?(@.kind=='impl' && @.inner.for.inner.name=='S')]"
```

doesn't work due to https://github.com/freestrings/jsonpath/issues/91

The way I to work around this is to add a doc comment to the impl and select on that, eg

```rust
pub struct S;

//// impl S
impl S {
    pub fn is_present() {}
}

// @set impl_S = "$.index[*][?(@.docs=='impl S')].id"
// @set is_present = "$.index[*][?(@.name=='is_present')].id"
// @is "$.index[*][?(@.name=='S')].inner.impls[*]" $impl_S
// @is "$.index[*][?(@.docs=='impl S')].inner.items[*]" $is_present
```

Comment (GuillaumeGomez) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs]: Smart trick! That doesn't allow to see how many impl blocks are implemented on `S` unfortunately but it's much better than currently. :)

**Comment by GuillaumeGomez** [comment]
Updated! Thanks for the suggestion @aDotInTheVoid !

File: src/test/rustdoc-json/reexport_method_from_private_module.rs (hunk 2 above)
Comment (aDotInTheVoid) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs]: If theirs a second impls on `S`, then it'll fail because several things match to `// @is "$.index[*][?(@.name=='S')].inner.impls[*]" $impl_S`, but because this in `#[no_core]`, their are no auto traits.

Similarly, this checks theirs only 1 method in the impl.

Comment (GuillaumeGomez) [reply src/test/rustdoc-json/reexport_method_from_private_module.rs]: Makes sense, thanks for the explanation! Then I guess it's ready for final review. :)
